kale/embed/video_se_cnn.py

```diff
@@ -0,0 +1,126 @@
+        y = self.avg_pool(x).view(b, c)
+        y = self.fc(y).view(b, c, 1, 1, 1)
+        # out = x * y.expand_as(x)
+        y = y - 0.5
+        out = x + x * y.expand_as(x)
+        return out
+
+
+class SELayerT(nn.Module):
+    def __init__(self, channel, reduction=2):
+        super(SELayerT, self).__init__()
+        self.avg_pool = nn.AdaptiveAvgPool3d(1)
+        self.fc = nn.Sequential(
+            nn.Linear(channel, channel // reduction, bias=False),
+            nn.ReLU(inplace=True),
+            nn.Linear(channel // reduction, channel, bias=False),
+            nn.Sigmoid(),
+        )
+
+    def forward(self, x):
+        b, _, t, _, _ = x.size()
+        output = x.transpose(1, 2).contiguous()
+        y = self.avg_pool(output).view(b, t)
+        y = self.fc(y).view(b, t, 1, 1, 1)
+        y = y.transpose(1, 2).contiguous()
+        # out = x * y.expand_as(x)
+        y = y - 0.5
+        out = x + x * y.expand_as(x)
+        return out
+
+
+class SELayerCoC(nn.Module):
+    def __init__(self, channel, reduction=16):
+        super(SELayerCoC, self).__init__()
+        self.conv1 = nn.Conv3d(in_channels=channel, out_channels=channel // reduction, kernel_size=1, bias=False)
+        self.bn1 = nn.BatchNorm3d(num_features=channel // reduction)
+        # self.conv3 = nn.Conv2d(
+        #     in_channels=channel // reduction,
+        #     out_channels=channel // reduction,
+        #     kernel_size=3,
+        #     padding=1,
+        #     groups=channel // reduction,
+        #     bias=False)
+        self.avg_pool = nn.AdaptiveAvgPool3d(1)
+        self.sigmoid = nn.Sigmoid()
+        self.conv2 = nn.Conv3d(in_channels=channel // reduction, out_channels=channel, kernel_size=1, bias=False)
+        self.bn2 = nn.BatchNorm3d(num_features=channel)
+
+    def forward(self, x):
+        b, c, t, _, _ = x.size()  # n, c, t, h, w
+        y = self.conv1(x)  # n, c/r, t, h, w
+        y = self.bn1(y)  # n, c/r, t, h, w
+        y = self.avg_pool(y)  # n, c/r, 1, 1, 1
+        y = self.conv2(y)  # n, c, 1, 1, 1
+        y = self.bn2(y)  # n, c, 1, 1, 1
+        y = self.sigmoid(y)  # n, c, 1, 1, 1
+        # out = x * y.expand_as(x)  # n, c, t, h, w
+        y = y - 0.5
+        out = x + x * y.expand_as(x)
+        return out
+
+
+class SELayerMC(nn.Module):
+    def __init__(self, channel, reduction=16):
+        super(SELayerMC, self).__init__()
+        self.max_pool = nn.AdaptiveMaxPool3d(1)
+        self.fc = nn.Sequential(
+            nn.Linear(channel, channel // reduction, bias=False),
+            nn.ReLU(inplace=True),
+            nn.Linear(channel // reduction, channel, bias=False),
+            nn.Sigmoid(),
+        )
+
+    def forward(self, x):
+        b, c, _, _, _ = x.size()
+        y = self.max_pool(x).view(b, c)
+        y = self.fc(y).view(b, c, 1, 1, 1)
+        # out = x * y.expand_as(x)
+        y = y - 0.5
+        out = x + x * y.expand_as(x)
+        return out
+
+
+class SELayerMAC(nn.Module):
```

Comment:
These classes are quite similar with quite some repetitions so maybe the code can be factored into more compact and reusable form.

Reply:
I have no idea how to do this. I am still exploring other SELayers and keeping them separately is easier for me to change and test. How about only keeping the basic layers and deleting others?

Reply:
Follow the principles in the PyKale video, put the above side to side and identify the redundancies to reduce, e.g. via introduce some structure or define some common functions. Ada did very well in their code. For the above, 
The init of SELayerMC and SELayerC are identical (and that of SELayerT is identical except a different default value for reduction) so you may make one parent or construct a common parent. It helps you to be more modular as well.